File: golang/sorting/quick/quick.go
package quick

func Sort(list []int) {
	SortHelper(list, 0, len(list) - 1)
}

func SortHelper(list []int, left int, right int) {
	if len(list) == 0 {
		return
	}

	index := partition(list, left, right)

	if left < index - 1 {
		SortHelper(list, left, index - 1)
	}

	if index < right {
		SortHelper(list, index, right)
	}

	return
}

func partition(list []int, left int, right int) int {
	pivot := list[(left + right) / 2]

	for left <= right {
		for list[left] < pivot {
			left++
		}

		for list[right] > pivot {
			right--
		}

		if left <= right {
			list[left], list[right] = list[right], list[left]
			left++
			right--
		}
	}
	return left
}

File: sorting/quick.py

def sort(lst, left=None, right=None):
    if not lst:
        return lst

    left = left or 0
    right = right or len(lst) - 1

    index = partition(lst, left, right)

    if left < index - 1:
        sort(lst, left, index - 1)

    if index < right:
        sort(lst, index, right)

    return lst


def partition(lst, left, right):
    pivot = lst[(left + right) // 2]

    while left <= right:
        while lst[left] < pivot:
            left += 1

        while lst[right] > pivot:
            right -= 1

        if left <= right:
            lst[left], lst[right] = lst[right], lst[left]
            left += 1
            right -= 1

    return left
